File: Binary Search/smallestDivisor.js
const smallestDivisor = (nums, threshold) => {
  const check = (divisor) => {
    let sum = 0;
    for (let num of nums) {
      sum += Math.ceil(num / divisor);
    }

    return sum <= threshold;
  }
  let left = 1, right = 10 ** 6;
  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (check(mid)) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
}

const nums = [1, 2, 5, 9],
  threshold = 6;

console.log(smallestDivisor(nums, threshold));

File: Binary Search/SmallestDivisor.java
public class SmallestDivisor {
  public int smallestDivisor(int[] nums, int threshold) {
    int left = 1;
    int right = 0;
    for (int num : nums)
      right = Math.max(num, right);

    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (check(mid, nums, threshold)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  public boolean check(double divisor, int[] nums, int threshold) {
    int sum = 0;
    for (double num : nums) {
      sum += Math.ceil(num / divisor);
    }

    return sum <= threshold;
  }
}
